### ISSUE TITLE
False positive for django-types Model.objects in strange circumstances

### ISSUE THREAD

**Issue by last-partizan** [microsoft/pyright#3542]
Hello!

After upgrading from `1.1.249` to `1.1.251` i started to see strange new bugs in my CI.

One of my models was missing `objects` attribute, which is typically would be set from metaclass.

I tried to create minimal reproduction, and it boils down to following example:

#### Test requirements:
```
pip install Django==4.0.5 django-types==0.15.0
```

#### Test files
```python
# example/models.py
from __future__ import annotations

from django.db import models


class Vps(models.Model):
    template = models.ForeignKey["OSTemplate"]("OSTemplate", models.CASCADE)


class ConfigTemplate(models.Model):
    pass


class OSTemplate(ConfigTemplate):
    pass


class VPSTemplate(ConfigTemplate):
    pass


OSTemplate.objects.get()
```

```python
# example/admin.py
from . import models


models.VPSTemplate.objects.all()
```

- Names must be exactly like this. If `admin.py` renamed to `utils.py` - bug vanishes.
- You must run `pyright` or `pyright example`, so it will scan files in alphabetical order. `admin.py` must be scanned before `models.py`.

#### pyright reports following errors:
```
Found 3 source files
/home/serg/src/tests/test-pyright/example/models.py
  /home/serg/src/tests/test-pyright/example/models.py:7:35 - error: Type "OSTemplate" cannot be assigned to type variable "_M@ForeignKey"
    Type "OSTemplate" is incompatible with bound type "Model | None" for type variable "_M@ForeignKey"
      Type "OSTemplate" cannot be assigned to type "Model | None"
        "OSTemplate" is incompatible with "Model"
        Type cannot be assigned to type "None" (reportGeneralTypeIssues)
  /home/serg/src/tests/test-pyright/example/models.py:22:12 - error: Cannot access member "objects" for type "Type[OSTemplate]"
    Member "objects" is unknown (reportGeneralTypeIssues)
```

What strange - with this minimal example bug happens even in `pyright 1.1.249`. But in production on big codebase old version works fine.

**To Reproduce**

You could create files yourself, or just clone my example: https://github.com/last-partizan/pyright-examples

**Expected behavior**
It should work just fine.

If you move `Vps` model to the bottom, it will work again. If you try to access any other model from `admin.py`, it will work just fine.

**Comment by erictraut** [comment]
Thanks for the bug report. I'm able to repro. This appears to be related to a circular type dependency that pyright is not handling well. I'll investigate further.

**Comment by erictraut** [comment]
Whew, this was a challenging bug to track down.

In case you're interested in the details...
First, it's important to know that pyright is a "lazy evaluator", so it always evaluates the types of expressions only as needed. Once the type of an expression node is established, it is cached in memory so it doesn't need to be evaluated again. In theory, evaluated types (and any errors that are discovered during the evaluation of those types) should not depend on the order in which expressions are evaluated. In this particular case, there was a bug that affected the type evaluation behavior based on certain evaluation ordering.

Here's the circularity that causes this issue:

1. In `admin.py`, the expression `models.VPSTemplate` requires the evaluation of type `VPSTemplate`.
2. Evaluating `VPSTemplate` requires evaluation of `ConfigTemplate`, which is its base class.
3. Evaluating `ConfigTemplate` requires evaluating the expression `models.Model`, which requires evaluating the expression `models`.
4. Evaluating the type of `models` requires the code flow graph to be traversed to determine whether this expression is reachable and whether the symbol's type is narrowed along any code paths.
5. To determine whether `models` is reachable, the code flow engine needs to determine whether the call `models.ForeignKey["OSTemplate"](...)` is a `NoReturn` call. This requires it to evaluate the expression `models.ForeignKey["OSTemplate"]`, which in turn requires it to evaluate the type of `OSTemplate`.
6. Evaluating `OSTemplate` requires evaluation of `ConfigTemplate`, which is its base class.
7. `ConfigTemplate` is still in the process of being evaluated above in step 2, so its base classes have not yet been established. When building its MRO (method resolution order), the MRO is incorrectly computed because the base classes are incomplete. This, in turn, causes the MRO for `OSTemplate` to be incorrectly computed.

The bug is that the corrupted (partially-constructed) type of `OSTemplate` was cached in memory and used in subsequent type evaluation steps.

The fix is to detect this case and not add the corrupted type into the internal type cache.

This was difficult to repro because it depended on evaluation ordering. Small perturbations in evaluation ordering avoided the code path that resulted in this corruption. That explains why you could make the problem go away by changing the order in which files were analyzed.

This will be addressed in the next release.

Thanks again for filing the bug report.

**Comment by last-partizan** [comment]
Yes, i'm interested in the details!

Thank you for fixing it and for the explanation!

**Comment by erictraut** [comment]
We'll leave the bug open until the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.253, which I just published. It will also be included in the next release of pylance.

**Comment by last-partizan** [comment]
Reporting back after updating to `1.1.253`, it fixed some cases, but if found one more!

Now we need to add `example/tests.py`, and a method to `Vps` model.


```diff
From 1f632bbb0579a3f82e67f8ddb8c7bc8d854444f0 Mon Sep 17 00:00:00 2001
From: Serg Tereshchenko <serg.partizan@gmail.com>
Date: Wed, 8 Jun 2022 13:06:05 +0300
Subject: [PATCH] feat: Add another failing case

---
 example/models.py | 3 +++
 example/tests.py  | 4 ++++
 2 files changed, 7 insertions(+)
 create mode 100644 example/tests.py

diff --git a/example/models.py b/example/models.py
index 67f5f66..412a955 100644
--- a/example/models.py
+++ b/example/models.py
@@ -6,6 +6,9 @@ from django.db import models
 class Vps(models.Model):
     template = models.ForeignKey["OSTemplate"]("OSTemplate", models.CASCADE)

+    def reinstall(self, template: OSTemplate):
+        ...
+

 class ConfigTemplate(models.Model):
     pass
diff --git a/example/tests.py b/example/tests.py
new file mode 100644
index 0000000..9f446aa
--- /dev/null
+++ b/example/tests.py
@@ -0,0 +1,4 @@
+from .models import Vps
+
+def test_reinstall(vps: Vps):
+    vps.reinstall(vps.template)
--
```

I pushed those changes to https://github.com/last-partizan/pyright-examples.


And it fails with following error:

```
/home/serg/src/tests/test-pyright/example/tests.py
  /home/serg/src/tests/test-pyright/example/tests.py:4:19 - error: Argument of type "OSTemplate" cannot be assigned to parameter "template" of type "OSTemplate" in function "reinstall"
    "example.models.OSTemplate" is incompatible with "example.models.OSTemplate" (reportGeneralTypeIssues)
```

**Comment by erictraut** [comment]
If you think there is still a bug, please open a new issue with a self-contained code sample.

**Comment by finviman** [comment]
just curiously, pyright can not find objects of a Model and it's methods if I don't install django-types.

**Comment by pabluez** [comment]
having the same problem...
![image](https://github.com/microsoft/pyright/assets/777052/be4022bb-dd68-4f65-a919-a5dde5fe235c)

**Comment by marekpiskorek** [comment]
@pabluez installing [django-stubs](https://github.com/typeddjango/django-stubs) solves this.
I see this as more of a Django-introduced problem than particularly pyright's fault.
